cpp/include/Ice/ServerAuthenticationOptions.h

```diff
@@ -0,0 +1,161 @@
+#        undef SECURITY_KERNEL
+#    endif
+#    define SECURITY_WIN32 1
+#    include <schannel.h>
+#    include <security.h>
+#    include <sspi.h>
+#    undef SECURITY_WIN32
+#elif defined(__APPLE__)
+#    include <Security/SecureTransport.h>
+#    include <Security/Security.h>
+#else
+#    include <openssl/ssl.h>
+#endif
+
+namespace Ice::SSL
+{
+    /**
+     * The SSL configuration properties for client connections.
+     */
+    struct ServerAuthenticationOptions
+    {
+#if defined(_WIN32)
+        /**
+         * The server credentials.
+         *
+         * [See Schannel
+         * Credentials](https://learn.microsoft.com/en-us/windows/win32/api/schannel/ns-schannel-sch_credentials).
+         */
+        CredHandle serverCredentials;
+
+        /**
+         * A callback that allows selecting the server credentials based on the server's host name. When the
+         * callback is set it has preference over the credentials set in serverCredentials.
+         *
+         * @param host The target server's host name.
+         * @return The server credentials.
+         */
+        std::function<CredHandle(const std::string& host)> serverCredentialsSelectionCallback;
+
+        /**
+         * A value indicating whether or not the server sends a client certificate request to the client.
+         */
+        bool clientCertificateRequired;
+
+        /**
+         * A callback that allows to manually validate the client certificate. When the callback is not provided the
+         * client certificate will be validated using the platform default validation mechanism.
+         *
+         * The validation callback should return true if the certificate chain is valid, false otherwise. Alternatively
+         * an exception can be throw to indicate an error.
+         *
+         * @param context A security context is an opaque data structure that contains security data relevant to the
+         * current connection.
+         * @param info The connection information.
+         * @return true if the certificate is valid, false otherwise.
+         *
+         * [See Manually Validating Schannel
+         * Credentials](https://learn.microsoft.com/en-us/windows/win32/secauthn/manually-validating-schannel-credentials).
+         */
+        std::function<bool(CtxtHandle context, const IceSSL::ConnectionInfoPtr& info)>
+            clientCertificateValidationCallback;
+#elif defined(__APPLE__)
+        /**
+         * The server's certificate chain.
+         */
+        CFArrayRef serverCertificateChain;
+
+        /**
+         * A callback that allows selecting a certificate chain based on the server's host name. When the callback is
+         * set it has preference over a certificate chain set in serverCertificateChain.
+         *
+         * @param host The server's host name.
+         * @return The server's certificate chain.
+         */
+        std::function<CFArrayRef(const std::string& host)> serverCertificateSelectionCallback;
```

Comment:
This callback allows selecting the certificate before SSL handshake starts, host parameter should probably be the adapter name, a bit more convenient way when you want to be able to upgrade the certificate without restart.